### PR TITLE
Added optional NonexistentType cache for type cache. 

### PR DESCRIPTION
**Pull request by qshen2000**
use flag --enable-nonexistent-type-cache to enable this feature.

### CODE REVIEW

File: flags.go
Comment (raj-prince) [inline flags.go:265]: You also need to change [here](https://github.com/GoogleCloudPlatform/gcsfuse/blob/1a2d9d6702fc82d19ac13f1c82dbe26094a86b5a/tools/mount_gcsfuse/main.go#L83) to support this flag while mounting through /etc/fs.

File: internal/fs/inode/dir_test.go
Comment (raj-prince) [inline internal/fs/inode/dir_test.go:497]: Can we also add the test-cases when both implicit-directory and Nonexistent type are true?

**Review by raj-prince — CHANGES_REQUESTED** [review]
Overall looks good. Added a couple of comments, please incorporate.

File: flags.go
Comment (sethiay) [inline flags.go:393]: Can we rename EnableNonexistentType to EnableNonexistentTypeCache here and everywhere?

Comment (qshen2000) [reply flags.go:393]: will do. Thanks for the feedback.

Comment (qshen2000) [reply flags.go]: Done!

File: internal/fs/inode/dir_test.go
Comment (qshen2000) [reply internal/fs/inode/dir_test.go:497]: Added.

File: flags.go
Comment (qshen2000) [reply flags.go:265]: Added!

**Comment by qshen2000** [comment]
> Overall looks good. Added a couple of comments, please incorporate.

Please review again. I have addressed all comments. Thanks!

**Review by raj-prince — COMMENTED** [review]
Two things:
(1) Integration test is failing.
(2) Let's assume we pass this flag with ttl as 10 minutes, and we create the same file/node in mean time using the same mount. Since, we are not refreshing the cash, it will still return nil. If this is the case, can we update the flag description to be vocal about this scenario.

**Comment by qshen2000** [comment]
> Two things: (1) Integration test is failing. (2) Let's assume we pass this flag with ttl as 10 minutes, and we create the same file/node in mean time using the same mount. Since, we are not refreshing the cash, it will still return nil. If this is the case, can we update the flag description to be vocal about this scenario.

Hello, 
I have updated the description for the new flag. The integration test error might not be related to my change and seems fixed after I re-sync to base. Please review!

**Review by raj-prince — APPROVED** [review]

**Comment by raj-prince** [comment]
Hi qshen2000@, can you please re-sync this branch again to base? We have recently made fix for this linux-text stuck issue (linux-tests cancelled after 10 minutes).

**Comment by qshen2000** [comment]
> Hi qshen2000@, can you please re-sync this branch again to base? We have recently made fix for this linux-test stuck issue (linux-tests cancelled after 10 minutes).

Done! Fingers crossed!

File: flags.go
Comment (vadlakondaswetha) [inline flags.go:273]: type: cache

File: internal/fs/fs.go
Comment (vadlakondaswetha) [inline internal/fs/fs.go:76]: file/directory.

File: tools/mount_gcsfuse/main.go
Comment (raj-prince) [inline tools/mount_gcsfuse/main.go:88]: We need to change the hyphen with underscore in this flag. Since, while mounting through fstab entry we pass the flag as underscore.

For example,
fstab entry: my-bucket /mount/point gcsfuse rw,noauto,user,key_file=/path/to/key.json

As you can see, we have provided the flag (key_file) with underscore not hyphen.

**Review by raj-prince — CHANGES_REQUESTED** [review]
Please incorporate.

File: tools/mount_gcsfuse/main.go
Comment (qshen2000) [reply tools/mount_gcsfuse/main.go:88]: I see. Thanks for the hints!

File: flags.go
Comment (qshen2000) [reply flags.go:273]: will do!

File: internal/fs/fs.go
Comment (qshen2000) [reply internal/fs/fs.go:76]: thanks!

**Review by raj-prince — APPROVED** [review]

**Comment by qshen2000** [comment]
Hello,

Can someone please help to merge this pull request or I have to resolve conflicts first myself?

**Comment by raj-prince** [comment]
Hi qshen2000, 

I have resolved the conflicts through browser only, can you please correct the indentation came due to last merge commit.

We can merge this PR, waiting for vadlakondaswetha@ approval.

-Prince

**Review by vadlakondaswetha — COMMENTED** [review]
Will be merged after lint issue is fixed

**Comment by qshen2000** [comment]
> Will be merged after lint issue is fixed

could you please help resolve the conflicts again? I think the lint error is due to the code is behind the main line now. Once we resolve the conflicts and re-sync, the lint error should be gone.

**Comment by qshen2000** [comment]
Thanks for helping resolve the conflicts. I have also fixed the indentation. Please help to merge vadlakondaswetha@.

**Review by vadlakondaswetha — APPROVED** [review]

**Comment by qshen2000** [comment]
Hello,

What would trigger a new release? Wondering when we can use the new flag in a released version. 

Thanks!

**Comment by vadlakondaswetha** [comment]
GCSFuse releases are done once a month. This changes should be live in the March end release.

**Pull request merged**
